File: core0/us/gov/cash/t_t.cpp
//===-                           S C R I P T  T.V.
//===-                           https://script.tv
//===-
//===-            Copyright (C) 2017-2024 Script Network
//===-            Copyright (C) 2017-2024 manicberet@gmail.com
//===-
//===-                      GNU GENERAL PUBLIC LICENSE
//===-                       Version 3, 29 June 2007
//===-
//===-    This program is free software: you can redistribute it and/or modify
//===-    it under the terms of the GPLv3 License as published by the Free
//===-    Software Foundation.
//===-
//===-    This program is distributed in the hope that it will be useful,
//===-    but WITHOUT ANY WARRANTY; without even the implied warranty of
//===-    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
//===-
//===-    You should have received a copy of the General Public License
//===-    along with this program, see LICENCE file.
//===-    see https://www.gnu.org/licenses
//===-
//===----------------------------------------------------------------------------
//===-
#include "accounts_t.h"
#include "types.h"
#include "t_t.h"

#define loglevel "gov/cash"
#define logclass "t_t"
#include <us/gov/logs.inc>

using namespace us::gov::cash;
using c = us::gov::cash::t_t;

const char* c::msg_sealed{". Sealed."};

c::t_t(): flags(0) {
}

c::t_t(const t_t& other) {
    for (auto& i: other) {
        emplace(i);
    }
    flags = other.flags;
}

void c::hash_data_to_sign(sigmsg_hasher_t& h) const {
    for (auto& i: *this) {
        h.write(i.first);
        h.write(i.second);
    }
    h.write(flags);
}

void c::hash_data(hasher_t& h) const {
    for (auto& i: *this) {
        h.write(i.first);
        h.write(i.second);
    }
    h.write(flags);
}

void c::on_destroy(accounts_t& ledger) {
    for(auto& i: *this) {
        auto j = ledger.find(i.first);
        if (j != ledger.end()) {
            j->second.box.add(i.first, i.second);
        }
    }
}

void c::merge(const c& other) {
    for (auto i = other.begin(); i != other.end(); ++i) {
        auto j = emplace(*i);
        if (!j.second) {
            j.first->second += i->second;
        }
    }
    if (other.is_sealed()) flags = 1; //if any is sealed, the result is sealed.
}

/*
1 address = 1 coin
t_t is the token container
*/

ko c::set_supply(const hash_t& token, const cash_t& supply) {
    log("set_supply.", token, supply);
    if (is_sealed()) {
        auto r = "KO 89785 Mint is sealed. Ignored attempt to set supply.";
        log(r);
        return r;
    }
    auto i = find(token);
    if (i == end()) {
        if (supply == -1) {
            emplace(token, 0);
            flags = 1;
            return ok;
        }
        if (supply <= 0) {
            auto r = "KO 89786 Supply must be positive integer.";
            log(r);
            return r;
        }
        emplace(token, supply);
        return ok;
    }
    log("cur value", i->second);
    if (supply == -1) { //seal mint. Makes current amount in circulation immutable, becoming a zero inflation coin.
        log("mint sealed.", token);
        flags = 1;
        return ok; // don't update the balance
    }
    if (supply == 0) {
        if (is_unsealed()) { //if sealed the item doesnt go away 
            log("delete entry");
            erase(i);
        }
        return ok;
    }
    if (supply < 0) {
        auto r = "KO 89787 Supply must be positive integer.";
        log(r);
        return r;
    }
    i->second = supply;
    return ok;
}

cash_t c::get_valueX(const hash_t& coin) const {
    auto i = find(coin);
    if (i == end()) {
        return 0;  //non existent mint
    }
    return i->second;
}

bool c::burn(const hash_t& addr, const hash_t& coin, const cash_t& amount) {
    assert(coin.is_not_zero());
    auto i = find(coin);
    if (i == end()) return false;
    if (i->second >= amount) {
        i->second -= amount;
        if (i->second == 0) {
            if (addr == coin && is_unsealed()) {
                erase(i);
            }
        }
        return true;
    }
    return false;
}

void c::dump(const string& prefix, const hash_t& addr, ostream& os) const {
    for (auto& i: *this) {
        os << prefix << "coin " << i.first << ' ' << i.second;
        if (addr == i.first) {
            os << ". Mint";
            if (is_sealed()) {
                os << msg_sealed;
            }
        }
        os << '\n';
    }
}

size_t c::blob_size() const {
    return b::blob_size() + blob_writer_t::blob_size(flags);
}

void c::to_blob(blob_writer_t& writer) const {
    log("to_blob", "cur", (uint64_t)(writer.cur - writer.blob.data()));
    b::to_blob(writer);
    writer.write(flags);
}

ko c::from_blob(blob_reader_t& reader) {
    log("from_blob", "cur", (uint64_t)(reader.cur - reader.blob.data()));
    auto r = b::from_blob(reader);
    if (is_ko(r)) {
        return r;
    }
    // <orma backwards_compatibility_hack version_from=0 version_to=11/>
    log("reader.header.version", +reader.header.version);
    if (reader.header.version < 11) {
        log("old version. flags set to 0");
        flags = 0; //unsealed
    }
    else {
        auto r = reader.read(flags);
        if (is_ko(r)) return r;
    }
    return ok;
}

File: core0/us/sdk/wallet/java/us/gov/cash/t_t.java
//===-                           S C R I P T  T.V.
//===-                           https://script.tv
//===-
//===-            Copyright (C) 2017-2024 Script Network
//===-            Copyright (C) 2017-2024 manicberet@gmail.com
//===-
//===-                      GNU GENERAL PUBLIC LICENSE
//===-                       Version 3, 29 June 2007
//===-
//===-    This program is free software: you can redistribute it and/or modify
//===-    it under the terms of the GPLv3 License as published by the Free
//===-    Software Foundation.
//===-
//===-    This program is distributed in the hope that it will be useful,
//===-    but WITHOUT ANY WARRANTY; without even the implied warranty of
//===-    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
//===-
//===-    You should have received a copy of the General Public License
//===-    along with this program, see LICENCE file.
//===-    see https://www.gnu.org/licenses
//===-
//===----------------------------------------------------------------------------
//===-
package us.gov.cash;
import us.gov.io.blob_reader_t;                                                                // blob_reader_t
import us.gov.io.blob_writer_t;                                                                // blob_writer_t
import static us.gov.cash.types.cash_t;                                                        // cash_t
import us.CFG;                                                                                 // CFG
import us.gov.crypto.ripemd160.hash_t;                                                         // hash_t
import static us.stdint.*;                                                                     // *
import us.ko;                                                                                  // ko
import java.util.Map;                                                                          // Map
import java.io.PrintStream;                                                                    // PrintStream
import us.gov.io.seriable;                                                                     // seriable
import us.gov.io.seriable_map__hash_cash;                                                      // seriable_map__hash_cash
import static us.gov.io.types.blob_t.serid_t;                                                  // serid_t

@SuppressWarnings("serial")
public class t_t extends seriable_map__hash_cash {

    static final String msg_sealed = ". Sealed.";

    public t_t() {
    }

    public t_t(final t_t other) {
        for (Map.Entry<hash_t, cash_t> i: other.entrySet()) {
            put(i.getKey(), i.getValue());
        }
        flags.value = other.flags.value;
    }

    public final cash_t get_value(final hash_t coin) {
        cash_t i = get(coin);
        if (i == null) {
            return cash_t.zero_;  //non existent mint
        }
        return i;
    }

    public void dump(final String prefix, final hash_t addr, PrintStream os) {
        for (Map.Entry<hash_t, cash_t> i: entrySet()) {
            os.print(prefix + "coin " + i.getKey().encode() + ' ' + i.getValue().value);
            if (addr.equals(i.getKey())) {
                os.print(". Mint");
                if (is_sealed()) {
                    os.print(msg_sealed);
                }
            }
            os.println();
        }
    }

    final uint8_t get_flags() { return flags; }
    boolean is_sealed() { return flags.value == 1; }
    boolean is_unsealed() { return flags.value == 0; }

    public int mint_type(final hash_t address) { //0 not a mint; 1 open mint; 2 sealed mint
        if (get(address) == null) return 0;
        if (is_unsealed()) return 1;
        return 2;
    }

    @Override public serid_t serial_id() { return serid_t.no_header; }

    @Override public int blob_size() {
        return super.blob_size() + blob_writer_t.blob_size(flags);
    }

    @Override public void to_blob(blob_writer_t writer) {
        super.to_blob(writer);
        writer.write(flags);
    }

    @Override public ko from_blob(blob_reader_t reader) {
        {
            ko r = super.from_blob(reader);
            if (ko.is_ko(r)) return r;
        }
        {
            ko r = reader.read(flags);
            if (ko.is_ko(r)) return r;
        }
        return ko.ok;
    }

    uint8_t flags = new uint8_t((short)0);
};


